wallet-new/src/Cardano/Wallet/API/V1/Errors.hs

```diff
@@ -54,7 +58,7 @@ import           Cardano.Wallet.API.V1.Types (SyncPercentage, SyncProgress (..),
 -- TODO: change fields' types to actual Cardano core types, like `Coin` and `Address`
 data WalletError =
       NotEnoughMoney { weNeedMore :: !Int }
-    | OutputIsRedeem { weAddress :: !Text }
+    | OutputIsRedeem { weAddress :: !(V1 Core.Address) }
```

Comment:
Sum types! Now with partial funtions!! :imp: 

Ok if I point Jordan at that data type?

Reply:
yes please! do you mean to factor it out into individual exception types? that would be great

Reply:
No the idea was just to remove all the field accessors (like `weAddress` in the constructor about) because all these field accessors are partial functions.

These are particularly odious in error type, because they are being used in the error handling path which is less likely to get exercised during regular execution or testing. What happens when you get an error that you'd like to log, but the error handling code gets a pattern match failure? You've just lost your logging.

Reply:
@erikd It would be lovely if you guys could take a look at that type, but there are a bunch of pitfalls you folks should be aware of, as that type was created like that on purpose, as crazy as it might sound. Let me explain, hehe.

A `WalletError` is much more than a way to report an error in our backend, is also a particular structure used to expose nice error messages to API consumers. In particular, we are using the [Jsend](https://labs.omniti.com/labs/jsend) format to expose errors to the clients. The pattern we settled upon with the frontend folks was to expose a tag (essentially the type constructor converted to a `String`) in the `message` field, and _extra_ diagnostics in the `diagnostic` field. You can look at how [this renders](https://cardanodocs.com/technical/wallet/api/v1/#section/Errors) in the relevant section of our API documentation. A practical example is the following: you want to express the output is not a redeem address (like on like 61). What we do is to show the frontend the following:

``` json
{
   "status":"error",
   "diagnostic":{
      "address":"b10b24203f1f0cadffcfd16277125cf7f3ad598983bef9123be80d93"
   },
   "message":"OutputIsRedeem"
}
```

See what we did there? We converted (via `generic-sop`) our type so that the `weAddress` is spliced back into the `address` field of our final JSON, and that's why we alas require partial functions.

By all means I'm very open to other solutions, provided we can maintain the same error semantic and the same output structure. I hope it clarifies the rationale 😉